ocaml/otherlibs/stdlib_alpha/capsule.ml

```diff
@@ -80,9 +105,67 @@ module Mutex = struct
       Password.unsafe_mk ()
 end
 
+module Rwlock = struct
+
+  type 'k t : value mod portable uncontended = { rwlock : Rw.t; mutable poisoned : bool }
+
+  type packed : value mod portable uncontended = P : 'k t -> packed
+
+  exception Poisoned
+
+  let with_write_lock :
+    'k t
+    -> ('k Password.t @ local -> 'a) @ local
+    -> 'a
+    @@ portable
+    = fun t f ->
+      Rw.lock_write t.rwlock;
+      match t.poisoned with
+      | true -> Rw.unlock t.rwlock; reraise Poisoned
```

Comment:
The `Mutex.with_lock` implementation makes a similar reraise if the poison flag is true, should both be changed, or just this one on line 124?

Reply:
Yeah, lines 124 and 141 should both be changed.

For context: When an exception is raised, the runtime collects a backtrace for the portion of stack being discarded, between the raise and the catch. If the handler for that exception uses `reraise` to continue propagating the exception, then the next piece of backtrace is _appended_ to the existing backtrace. This mechanism allows the backtrace for e.g. `with_write_lock lk (fun _ -> failwith "blah")` to point at the `failwith`, even though the exception was most recently raised by the handler logic inside `with_write_lock`.

This mechanism only makes sense when the runtime's global backtrace buffer contains something related to the exception you're about to reraise. In the case where `t.poisoned` is true, the backtrace buffer contains whatever happens to be the most recent backtrace is whatever code most recently raised. Appending to that backtrace is not going to give meaningful results.

In other words, `reraise` only really makes sense inside the exception-handling arm of a `try`, which describes some but not all of the uses of `reraise` here.

(If you really wanted to, you could use reraise here by capturing the backtrace of the point that failed when setting `poisoned <- true`, and restoring it using `Printexc.raise_with_backtrace`, to guarantee you're appending to the correct backtrace. I don't recommend doing this, but that's what correct uses of reraise outside of exception handlers have to look like)

Reply:
Interesting! And yes the explanation makes sense. In total I think there are 5 uses of `reraise` outside an exception-handling arm: lines 86, 101, 124 and 144 and 155. If I understand correctly, we could replace all those with a `raise`? 

Out of curiosity, what happens if an exception is re-raised via `reraise` in one thread (by catching a exception being thrown by some call-back function f), after which another thread then raises a `Poisoned` exception via `raise`, does each thread have its own backtrace?

Reply:
> If I understand correctly, we could replace all those with a raise?

Yes - I hadn't spotted those other ones!

> does each thread have its own backtrace?

Yes, in order to handle the situation you describe. (If you're curious, grep for `backtrace_buffer` in `ocaml/runtime` and `ocaml/otherlibs/systhreads`)

Reply:
I believe @stedolan has forgotten that on our branch `raise` is defined as `%reraise` for confusing semi-historical reasons. I think the right fix here is actually to rename the `external reraise ...` in this file to `external raise ...`

Reply:
I'm going to resolve this, since the confusing name for `raise` is unrelated to this PR.

Reply:
Nope, haven't forgotten that, and I still this this is a problem. Since we define `raise` as `reraise`, our backtraces are generally broken. When we fix this, the backtraces elsewhere will improve, but backtraces arising from poisoned mutexes will still be wrong due to this redefinition.

(I agree it's an instance of a thing that is generally broken elsewhere, but since it's being explicitly redefined here we may as well get it right this time)

Reply:
Hmm, I guess. `raise` is only being re-defined here because the main one isn't portable yet. As soon as it is I think we should call it normally here rather than redefine it. In which case its behaviour will go straight back to `%reraise`, so I don't think I see the benefit of using `%raise` here temporarily.

Reply:
If you feel strongly about it maybe change the existing definition in a separate PR? It seems unrelated to this PR.